### PR TITLE
Comprehensions - Week 7

### PR DESCRIPTION
**Pull request by YuxiaoRan**

### CODE REVIEW

**Comment by YuxiaoRan** [comment]
#### Status Update

Fixed some minor bugs in the program. For the most up-to-date version, see [here](https://github.com/YuxiaoRan/chocopy-wasm-compiler-B/tree/week7).

Now the compiler supports a comprehension like this:
```python
(<expr> for <id> in Range().new(<start>, <end>, <step>) if <cond>)
```
where `if <cond>` is an optional filter, and `<expr>` can be any expression -- even a ternary expression (`<exprIfTrue> if <someCond> else <exprIfFalse>`).

For example,
```python
g : generator = None
g = (a * 2 if a % 4 == 0 else a * 3 for a in Range().new(0, 20, 1) if a % 2 == 0)
```
will __print__ out all the values generated by the generator `g` (for now).

In the next milestone, rather than printing out the values, we will instead fill the data into the actual `list/set/dict/...`.

File: ast.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@ export type Type =
   | {tag: "none"}
   | {tag: "class", name: string}
   | {tag: "either", left: Type, right: Type }
+  | {tag: "generator", type: Type } // generator type
```
Comment (yalhessi) [inline ast.ts:10]: Your generator still has a type in the abstract syntax but not one in the concrete syntax.

File: ast.ts [hunk 2]
```diff
@@ -42,6 +53,8 @@ export type Expr<A> =
   | {  a?: A, tag: "index", obj: Expr<A>, index: Expr<A> }
   | {  a?: A, tag: "method-call", obj: Expr<A>, method: string, arguments: Array<Expr<A>> }
   | {  a?: A, tag: "construct", name: string }
+  | {  a?: A, tag: "comprehension", type: Type, lhs: Expr<A>, item: string, iterable: Expr<A>, ifcond?: Expr<A> } // comprehension expression
```
Comment (yalhessi) [inline ast.ts:56]: What's this `Type`? Can you have an `int` comprehension?

File: type-check.ts [hunk 3]
```diff
@@ -375,6 +417,47 @@ export function tcExpr(env : GlobalTypeEnv, locals : LocalTypeEnv, expr : Expr<S
...
+      const eitherTyp : Type = { tag: "either", left: exprIfTrueTyp, right: exprIfFalseTyp };
+      return { ...expr, a: [eitherTyp, expr.a], exprIfTrue: tExprIfTrue, ifcond: tIfCond, exprIfFalse: tExprIfFalse };
+    case "comprehension":
+      const tIterable = tcExpr(env, locals, expr.iterable);
+      const [isIterable, itemTyp] = isIterableType(tIterable.a[0])
+      if (!isIterable) {
+        throw new TypeCheckError(`Type ${tIterable.a[0]} is not iterable`);
+      }
+      if (locals.vars.has(expr.item)) {
+        throw new TypeCheckError(`id ${expr.item} already exists locally`);
```
Comment (yalhessi) [inline type-check.ts:441]: Is it necessary to reject an id if it's used elsewhere? Why can't you shadow the variable name?

File: type-check.ts [hunk 4]
```diff
@@ -52,25 +52,67 @@ export type TypeError = {
...
 }
 
+export function isIterableType(t: Type) : [Boolean, Type] {
+  // check if t is an iterable type
+  // if true, also return type of each item in the iterable
+  switch (t.tag) {
+    case "either":
+      return isIterableType(t.left) || isIterableType(t.right);
+    case "class":
+      return [t.name === "Range", NUM]; // for MS1, Range object is considered as iterable
```
Comment (yalhessi) [inline type-check.ts:90]: Why must a class name be `Range`?

File: repl.ts [hunk 5]
```diff
@@ -9,6 +9,44 @@ interface REPL {
...
+    self.max = max
+    self.stp = stp
+    return self
+  def next(self : Range)->int:
+    c : int = 0
+    c = self.cur
+    self.cur = self.cur + self.stp
+    return c
+  def hasnext(self : Range)->bool:
+    return self.cur < self.max
```
Comment (yalhessi) [inline repl.ts:30]: What if the step is negative?

File: repl.ts [hunk 6]
```diff
@@ -34,8 +72,15 @@ export class BasicREPL {
...
+
+    // check if built-in classes are parsed
+    var sourceCode = source;
+    if (!this.currentEnv.classes.has("Range") || !this.currentEnv.classes.has("generator")) {
+      sourceCode = builtinClasses + source;
+    }
+
     const config : Config = {importObject: this.importObject, env: this.currentEnv, typeEnv: this.currentTypeEnv, functions: this.functions};
-    const [result, newEnv, newTypeEnv, newFunctions, instance] = await run(source, config);
+    const [result, newEnv, newTypeEnv, newFunctions, instance] = await run(sourceCode, config);
```
Comment (yalhessi) [inline repl.ts:83]: This is okay for now, but we would not want to be appending to the source code before the user runs their code. If there are errors here, what would we report to the user?

**Review by yalhessi — CHANGES_REQUESTED** [review]
Good work!

I was more focused on your implementation here in my review than your design, but to be honest, it's hard to piece together what's going on. I have no clue which of your tests you got working, and I think you have too many unfinished pieces in here, so here are some suggestions that will hopefully help you make progress:
1) Pull your tests into a test file so you can run them. This is also a quick way for us to tell how much is working.

2) Maybe for now focus on a much more limited syntax. I know comprehensions have a lot of coverage, but I think you're trending in the right direction. For now, maybe just try to get a generator with a range object inside of it working. I think you have a reasonable implementation for `range` and `generator` but you should think of these two differently. A user can write a different iterable that also implements `next` and `hasnext` so your implementation should be flexible there. However, you can treat `generator` as a builtin since you have full control over the syntax and semantics of generators.

**Comment by yalhessi** [comment]
I would focus on getting an example like this working:
```
g1 : generator = None
g2 : generator = None
g1 = (a * 2 if a % 4 == 0 else a * 3 for a in Range().new(0, 20, 1) if a % 2 == 0)
g2 = (print(a) for a in g1)
```

I think this would help you tie together what you've been working on into a functioning example.

File: ast.ts (hunk 1 above)
Comment (YuxiaoRan) [reply ast.ts:10]: Could you please clarify what a concrete syntax means? `type` is the type of elements generated by this generator.
E.g. `(a for a in [1, 2, 3])` would be a generator with type `number`.

File: ast.ts (hunk 2 above)
Comment (YuxiaoRan) [reply ast.ts:56]: `type` here is the type of the comprehension (i.e. what type of object the comprehension will return). E.g. `[a for a in [1, 2, 3]]` has type `list` (assuming `list` is a valid type implemented by the list group), while `(a for a in [1, 2, 3])` has type `generator`.

File: type-check.ts (hunk 3 above)
Comment (YuxiaoRan) [reply type-check.ts:441]: We'll make this change. Thanks.

File: type-check.ts (hunk 4 above)
Comment (YuxiaoRan) [reply type-check.ts:90]: This is just for MS1. In the next milestone, we will make this to accommodate any class that has `next()` and `hasnext()`.

File: repl.ts (hunk 5 above)
Comment (YuxiaoRan) [reply repl.ts:30]: We'll use the range function implemented by the for-loop group, but yes I will definitely consider this case when merging.

File: type-check.ts (hunk 4 above)
Comment (YuxiaoRan) [reply type-check.ts:90]: Resolved. Now `case "class"` checks if the class has `next()` and `hasnext()`. Also added cases for `list` and `set`.

File: type-check.ts (hunk 3 above)
Comment (YuxiaoRan) [reply type-check.ts:441]: Resolved. Now iterator name is always shadowed. E.g.
```python
(a + 1 for a in ...)
(a * 2 for a in ...)
```
will generate temporary comprehension variables `compvar$a1` and `compvar$a2` in the WASM code.

File: repl.ts (hunk 6 above)
Comment (YuxiaoRan) [reply repl.ts:83]: What would be an optimal way to add built-in function/classes to the program? In the future, should we append built-in code at the WASM level (e.g. inside `wasmSource` from `runner.ts`)?

**Pull request merged**